### ISSUE TITLE
[Feature Request] More command line workflow tools: build & run to simulator or device

### ISSUE THREAD

**Issue by colinta** [fastlane/fastlane#1062]
Now that I'm using Fastlane, I'm very much reminded of the command-line tools that I enjoy when developing apps using RubyMotion.  Programming languages aside, one of the best features of RubyMotion is the Rakefile-based build process.  Running `rake`, by default, compiles the app and runs it in the simulator.  You can specify which simulator to run (`rake device_family="iPhone 6s"`) or install and run to your device (`rake device`).

At the moment, I think we are hindered in our ability to implement this feature, because compiling xcode projects (using `xcodebuild` or `xctool`) does a **full** compilation, no iterative building support there.  Iterative builds is a must have, I think, for this kind of feature.  So we'll have to keep an eye on the build tools, or investigate them to find out if iterative builds *are* possible from the command line.

The net result here would be an amazing new option for developers who are *ahem* less than happy using Xcode.  For instance, using RubyMotion, I *rarely if ever* opened Xcode (I eschew XIBs/Storyboards), and life was grand.  Just me, Sublime Text, and a console, thank you very much.

# Outline

```
fastlane sim  # compiles and installs to sim
fastlane sim device="iPhone 6"
fastlane device  # compiles and installs to device
```

# Useful commands

I attempted this in the past, but the lack of iterative builds prevented it from being pragmatic

```
xcrun simctl install $device $app

device = look in ~/Library/Developer/CoreSimulator/Devices, find the one with a device.plist that matches the desired device
app = need to look in the build folder, e.g. Build/Products/Debug-iphonesimulator/Your.app

xcrun simctl launch $device $app_id

app_id = the reverse DNS name, com.your.App
```

**Comment by jcampbell05** [comment]
So there are 2 parts to this:

- The commands, could be implemented via lanes to trigger those commands. I think fastlane supports default lanes too.
- The second issue of iterative builds should be handled by the action itself unless this is an easy flag to expose in fastlane, then this is out of scope and a special tool should be built for this.

**Comment by colinta** [comment]
Yup seems reasonable.  I mentioned this feature to @KrauseFx and he asked me to open an issue so he/you can track it and learn more.  Not sure if he had ideas for it, or just wanted me to expound my idea.

**Comment by jcampbell05** [comment]
@colinta  It's good to hear new ideas :) There is lots to be worked on buildwise with Xcode :)

**Comment by KrauseFx** [comment]
Thanks for publishing this here. When you call fastlane you always run `fastlane [lane_name]`, so if we'd implement something like this, you'd have a fastlane action running your app in the simulator which can be added to any lane.

**Comment by KrauseFx** [comment]
Seems like the awesome @7mllm7 submitted this exact action with https://github.com/fastlane/fastlane/pull/1032 :+1:

**Comment by colinta** [comment]
Well, OK, I think familiarity with RubyMotion would help, because apparently I'm doing a poor job of emphasizing the incremental builds, which #1032 does not address (and, to be clear, I *don't think it's supported by the compiler*).

Having this feature be a serious addition to your workflow requires this; having to recompile the app from scratch every time is untenable.  I would rather have this feature *not implemented* than to have it checked off as complete, even though it's half baked.

**Comment by 7mllm7** [comment]
The main idea behind #1032 was actually __not__ to be used as part of "every day" development, but rather, like @KrauseFx [said](https://github.com/fastlane/fastlane/issues/1062#issuecomment-169573345), to be an action that can be added to fastlane lanes. We use it with our QA team, that only needs to test the app on different devices, and this way we can let drive a lane. This lane for instance pulls the recent changes, installs cocoapods, and does other actions that you can't expect QA guys to perform every time they want to manually test the app.

From my understanding, adding 'command line workflow tools' like this issue suggests, is a whole other topic.

**Comment by colinta** [comment]
Yup, I agree.  Awesome to have #1032 to build off, but as a daily workflow tool, we would need to investigate the compiler more.

**Comment by jcampbell05** [comment]
I think Xcode supports incremental builds by default. It will cache object files in the derived data folder. Often most tools will run a `clean` which forces xcode to clean that folder and rebuild everything. This is mainly because xcode isn't always great at detecting which object files to recompile. Something that built fine should actually fail and in fact I've had Xcode suddenly fail to compile because I ran a clean.

I suppose if this is a local build it should be okay but for CI  or releasing I strongly reccomen letting `gym` run a clean :)

**Comment by jcampbell05** [comment]
Maybe you should see if `gym` supports disabling the clean stage ? (if it has one). Does it @KrauseFx  ?

**Comment by colinta** [comment]
@jcampbell05 I'm experimenting, and *it does appear* that xcodebuild has added this – or maybe I was mistaken to begin with!  Last spring, I wrote a `rake` task to accomplish this same thing, and it was doing a clean compile every time.  That was in the days of Swift 1.1 and 30 minute compile times.  Let's all try to forget those dark days...

...anyway!  I updated the command a bit to bring it up-to-speed, and it seems to be on par with building via Xcode.  That's great!  So the only thing to check is whether the clean stage is automatic or not.
